File: leetcode/876.py
"""
Runtime: 28 ms, faster than 88.49% of Python3 online submissions for Middle of the Linked List.
Memory Usage: 14.2 MB, less than 72.92% of Python3 online submissions for Middle of the Linked List.
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr, n = head, 0
        while curr:
            curr = curr.next
            n += 1

        curr = head
        for i in range(n // 2):
            curr = curr.next

        return curr


"""
Runtime: 28 ms, faster than 87.15% of Python3 online submissions for Middle of the Linked List.
Memory Usage: 14.2 MB, less than 71.97% of Python3 online submissions for Middle of the Linked List.
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        count = 0
        curr = head
        while curr:
            count += 1
            curr = curr.next

        curr = head
        for i in range(count // 2):
            curr = curr.next

        return curr

File: leetcode/876.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Middle of the Linked List.
// Memory Usage: 7.1 MB, less than 21.54% of C++ online submissions for Middle of the Linked List.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode *middleNode(ListNode *head) {
    ListNode *slow = head, *fast = head;
    while (fast->next != nullptr && fast->next->next != nullptr) {
      slow = slow->next;
      fast = fast->next->next;
    }
    return fast->next == nullptr ? slow : slow->next;
  }
};
